emsdk

```diff
@@ -100,26 +100,46 @@ def which(program):
 
   return None
 
+def vswhere(version):
+  program_files = os.environ['ProgramFiles(x86)'] if 'ProgramFiles(x86)' in os.environ else os.environ['ProgramFiles']
+  vswhere_path = os.path.join(program_files, 'Microsoft Visual Studio', 'Installer', 'vswhere.exe')
+  output = json.loads(subprocess.check_output([vswhere_path, '-latest', '-version', '[%s.0,%s.0)' % (version, version + 1) , '-requires', 'Microsoft.VisualStudio.Component.VC.Tools.x86.x64', '-property', 'installationPath', '-format', 'json']))
+  return output[0]['installationPath'].encode('ascii') if len(output) > 0 else ''
+
+def vs_filewhere(installation_path, platform, file):
+  vcvarsall = os.path.join(installation_path, 'VC/Auxiliary/Build/vcvarsall.bat')
+  env = subprocess.check_output('cmd /c "%s" %s & where %s' % (vcvarsall, platform, file))
+  paths = [path[:-len(file)] for path in env.split('\r\n') if path.endswith(file)]
+  if (len(paths) > 0):
+    return paths[0]
+  else:
+    raise Exception('Cannot find the file in the given Visual Studio environment')
+
 CMAKE_GENERATOR = 'Unix Makefiles'
 if WINDOWS:
   # Detect which CMake generator to use when building on Windows
   if '--mingw' in sys.argv: CMAKE_GENERATOR = 'MinGW Makefiles'
   elif '--vs2013' in sys.argv: CMAKE_GENERATOR = 'Visual Studio 12'
   elif '--vs2015' in sys.argv: CMAKE_GENERATOR = 'Visual Studio 14'
+  elif '--vs2017' in sys.argv: CMAKE_GENERATOR = 'Visual Studio 15'
   else:
-    vs2015_exists = 'VS140COMNTOOLS' in os.environ or 'VSSDK140Install' in os.environ or os.path.isdir(os.path.join(os.environ['ProgramFiles'], 'Microsoft Visual Studio 14.0')) or ('ProgramFiles(x86)' in os.environ and os.path.isdir(os.path.join(os.environ['ProgramFiles(x86)'], 'Microsoft Visual Studio 14.0')))
-    vs2013_exists = 'VS120COMNTOOLS' in os.environ or os.path.isdir(os.path.join(os.environ['ProgramFiles'], 'Microsoft Visual Studio 12.0')) or ('ProgramFiles(x86)' in os.environ and os.path.isdir(os.path.join(os.environ['ProgramFiles(x86)'], 'Microsoft Visual Studio 12.0')))
+    program_files = os.environ['ProgramFiles(x86)'] if 'ProgramFiles(x86)' in os.environ else os.environ['ProgramFiles']
+    vs2017_exists = len(vswhere(15)) > 0
+    vs2015_exists = 'VS140COMNTOOLS' in os.environ or 'VSSDK140Install' in os.environ or os.path.isdir(os.path.join(program_files, 'Microsoft Visual Studio 14.0'))
+    vs2013_exists = 'VS120COMNTOOLS' in os.environ or os.path.isdir(os.path.join(program_files, 'Microsoft Visual Studio 12.0'))
     mingw_exists = which('mingw32-make') != None and which('g++') != None
-    if vs2015_exists: CMAKE_GENERATOR = 'Visual Studio 14'
+    if vs2017_exists: CMAKE_GENERATOR = 'Visual Studio 15'
+    elif vs2015_exists: CMAKE_GENERATOR = 'Visual Studio 14'
```

Comment:
Thanks, this looks like a good addition. You mention that currently build fails, so can you change the above two lines to take precedence to default to Visual Studio 2015 since otherwise this would choose the breaking VS2017 by default?